pkg/network/gvtap/gvtap.go

```diff
@@ -0,0 +1,492 @@
+	"io"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/containers/gvisor-tap-vsock/pkg/tap"
+	"github.com/containers/gvisor-tap-vsock/pkg/types"
+	"github.com/containers/gvisor-tap-vsock/pkg/virtualnetwork"
+	"github.com/sirupsen/logrus"
+	"github.com/songgao/water"
+
+	"github.com/rootless-containers/rootlesskit/v2/pkg/api"
+	"github.com/rootless-containers/rootlesskit/v2/pkg/common"
+	"github.com/rootless-containers/rootlesskit/v2/pkg/messages"
+	"github.com/rootless-containers/rootlesskit/v2/pkg/network"
+	"github.com/rootless-containers/rootlesskit/v2/pkg/network/iputils"
+	"github.com/rootless-containers/rootlesskit/v2/pkg/network/parentutils"
+)
+
+const (
+	DriverName = "gvisor-tap-vsock"
+	gatewayIp  = "10.0.2.1"
+)
+
+// NewParentDriver instantiates a new parent driver
+func NewParentDriver(logWriter io.Writer, mtu int, ipnet *net.IPNet, ifname string, disableHostLoopback bool, enableIPv6 bool) (network.ParentDriver, error) {
+	if mtu < 0 {
+		return nil, errors.New("got negative mtu")
+	}
+	if mtu == 0 {
+		mtu = 65520
+	}
+
+	if ifname == "" {
+		ifname = "tap0"
+	}
+
+	return &parentDriver{
+		logWriter:           logWriter,
+		mtu:                 mtu,
+		ipnet:               ipnet,
+		ifname:              ifname,
+		disableHostLoopback: disableHostLoopback,
+		enableIPv6:          enableIPv6,
+	}, nil
+}
+
+type parentDriver struct {
+	logWriter           io.Writer
+	mtu                 int
+	ipnet               *net.IPNet
+	ifname              string
+	disableHostLoopback bool
+	enableIPv6          bool
+	infoMu              sync.RWMutex
+	info                func() *api.NetworkDriverInfo
+
+	// Store the virtual network and its network switch for later use
+	vn            *virtualnetwork.VirtualNetwork
+	networkSwitch *tap.Switch
+	vnMu          sync.RWMutex
+
+	// Socket for communication with the child namespace
+	socketPath string
+	listener   net.Listener
+	ctx        context.Context
+	cancel     context.CancelFunc
+}
+
+func (d *parentDriver) Info(ctx context.Context) (*api.NetworkDriverInfo, error) {
+	d.infoMu.RLock()
+	infoFn := d.info
+	d.infoMu.RUnlock()
+	if infoFn == nil {
+		return &api.NetworkDriverInfo{
+			Driver: DriverName,
+		}, nil
+	}
+
+	return infoFn(), nil
+}
+
+func (d *parentDriver) MTU() int {
+	return d.mtu
+}
+
+// setupNetworkConfig sets up the basic network configuration
+func (d *parentDriver) setupNetworkConfig() (ip string, gateway string, netmask int, err error) {
+	if d.ipnet != nil {
+		// Use provided CIDR
+		var ipAddr net.IP
+		var gw net.IP
+
+		ipAddr, err = iputils.AddIPInt(d.ipnet.IP, 100)
+		if err != nil {
+			return "", "", 0, err
+		}
+		ip = ipAddr.String()
+
+		gw, err = iputils.AddIPInt(d.ipnet.IP, 1)
+		if err != nil {
+			return "", "", 0, err
+		}
+		gateway = gw.String()
+		netmask, _ = d.ipnet.Mask.Size()
+	} else {
+		// Use default network
+		ip = "10.0.2.100"
+		gateway = gatewayIp
+		netmask = 24
+	}
```

Comment:
Why do we need this `else` block?
`AddIPInt` should work for the default subnet too?